src/EnumDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,84 @@
+<?php
+
+declare(strict_types=1);
+
+namespace MabeEnumPHPStan;
+
+use MabeEnum\Enum;
+use PhpParser\Node\Expr\MethodCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\MethodReflection;
+use PHPStan\Reflection\ParametersAcceptorSelector;
+use PHPStan\ShouldNotHappenException;
+use PHPStan\Type\ArrayType;
+use PHPStan\Type\DynamicMethodReturnTypeExtension;
+use PHPStan\Type\IntegerType;
+use PHPStan\Type\Type;
+use PHPStan\Type\TypeCombinator;
+
+class EnumDynamicReturnTypeExtension implements DynamicMethodReturnTypeExtension
+{
+    /**
+     * Buffer of known return types of Enum::getValues()
+     * @var Type[]
+     * @phpstan-var array<class-string<Enum>, Type>
+     */
+    private $enumValuesTypeBuffer = [];
+
+    public function getClass(): string
+    {
+        return Enum::class;
+    }
+
+    public function isMethodSupported(MethodReflection $methodReflection): bool
+    {
+        $supportedMethods = ['getvalue'];
+        if (method_exists(Enum::class, 'getValues')) {
+            array_push($supportedMethods, 'getvalues');
+        }
+
+        return in_array(strtolower($methodReflection->getName()), $supportedMethods, true);
+    }
+
+    public function getTypeFromMethodCall(MethodReflection $methodReflection, MethodCall $methodCall, Scope $scope): Type
+    {
+        $enumType   = $scope->getType($methodCall->var);
+        $methodName    = $methodReflection->getName();
+        $methodClasses = $enumType->getReferencedClasses();
+        if (count($methodClasses) !== 1) {
+            return ParametersAcceptorSelector::selectSingle($methodReflection->getVariants())->getReturnType();
+        }
+
+        $enumeration = $methodClasses[0];
+
+        switch (strtolower($methodName)) {
+            case 'getvalue':
+                return $this->getEnumValuesType($enumeration, $scope);
+            case 'getvalues':
+                return new ArrayType(
```

Comment:
I would create ConstantArrayType if we really know all items (not generalized once).

Reply:
Yes this could be done